new/rfcXXX-telemetry.md

```diff
@@ -0,0 +1,94 @@
+---
+RFC: unassigned
+Author: Ranjib Dey <ranjib@linux.com>
+Status: Draft
+Type: Standards Track
+---
+
+# Title
+
+Chef telemetry
+
+## Motivation
+
+  As a chef user and developer
+  I want an easy way to get chef run performance data
+  so that I can identify areas that can be optimized
+
+As we automate more things with chef and control more system resources,
+chef run time increases, and the chef client process consumes more memory and cpu.
+As a configuration management system, it is important that a chef run
+does not impact the underlying host performance. Over the last few years
+Chef community has shared a number of optimization techniques. Aim of this
+RFC is to facilitate some of those common patterns by providing a robust
+telemetry system in core chef. If configured, the telemetry system will give users fine
+grained metrics about chef run that can be used to find out
+performance hiccups. This will also help users quantify how a chef
+run impacts the underlying system.
+
+## Specification
+
+Telemetry module will provide following metrics:
+
+- Time spent on key chef run milestones: How long does the chef run take for each stage? Telemetry system
+  will declare  a subset of the current events as major milestones, these are: run_start,
+  ohai_completed, node_load_completed, cookbook_resolution_complete, converge_start,
+  converge_complete, run_completed/failed. Metrics will be captured during these milestones via event handlers.
+  This can be used to estimate time spent on cookbook sync, compilation
+  convergence phase etc.
+
+- Time spent on individual resource & recipe convergence. This information is already available
+  using the API, telemetry system will just consolidate them, along side the other metrics.
+
+- GC related data. Ruby's internal memory usage obtained from
+  GC.stat, captured during the chef run milestones.
+
+Telemetry system will also let users define their custom metrics that can be computed
+during these milestones (like total chef client process memory consumption from /proc/PID),
+This will be offered via Chef class DSL. This will also be the recommended way to implement platform
+specific metrics. Example:
+
+```ruby
+require 'sys/proctable'
+
+Chef.telemetry do |telemetry|
+  telemetry.add_metric 'vsize' do
+    Sys::ProcTable.ps(Process.pid).vsize
+  end
+end
+```
+
+In addition to capturing these metrics, telemetry system will also allow publishing
+them via statsd or saving them as node attribute.
```

Comment:
Might we run into an issue here where we end up dealing with requests to add the ability to publish using other formats? Although commonly used, statsd isn't ubiquitous and if implementation of this is too tied to that format it might  hinder future expandability.